### PR TITLE
DAOS-10930 pydaos: Adding option to pass the object class for dictionary as an argument.

### PR DESCRIPTION
**Pull request by ravalsam**
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: docs/user/python.md [hunk 1]
```diff
@@ -74,6 +74,17 @@ stadium
 2
 ```
 
+User can pass the predefined object class id during dict() method call.
```
Comment (mchaarawi) [inline docs/user/python.md:77]: it would be good to expand this and mention passing the oclass is optional; can be RP or EC or S; and has to satisfy the rf property on the container. Also one could create the container with RF setting indicating that the oclass used automatically would use replication.
- in your example it might encourage using EC for dictionaries, which we should not do ;-); so please use RP instead.

File: src/client/pydaos/pydaos_core.py [hunk 2]
```diff
@@ -107,11 +107,12 @@ def get(self, name):
     def __getitem__(self, name):
         return self.get(name)
 
-    def dict(self, name, v: dict = None):
+    def dict(self, name, v: dict = None, cid="OC_SX"):
```
Comment (mchaarawi) [inline src/client/pydaos/pydaos_core.py:110]: why is this SX? is that default?
if yes, it should be 0

File: src/client/pydaos/pydaos_core.py [hunk 3]
```diff
@@ -124,12 +125,13 @@ def dict(self, name, v: dict = None):
 
         return dd
 
-    def array(self, name, v: list = None):
+    def array(self, name, v: list = None, cid="OC_SX"):
```
Comment (mchaarawi) [inline src/client/pydaos/pydaos_core.py:128]: same comment as dict

**Review by mchaarawi — CHANGES_REQUESTED** [review]
- please add some regression test for this (possibly update the pydaos test in NLT).
- to verify we probably need a get_info on object to query the object class and print for the user to make sure of the oclass being used. same as like daos fs get-attr for posix.

File: docs/user/python.md (hunk 1 above)
Comment (ravalsam) [reply docs/user/python.md:77]: I will extend the comments. Example has OC_RP_2G1 not EC, so it should be ok.

File: src/client/pydaos/pydaos_core.py (hunk 3 above)
Comment (ravalsam) [reply src/client/pydaos/pydaos_core.py:128]: Will use 0 as default

File: src/client/pydaos/pydaos_core.py [hunk 4]
```diff
@@ -107,12 +107,24 @@ def get(self, name):
...
 
-    def dict(self, name, v: dict = None):
+    def _get_object_id(self, cid):
+        """ Get the existing DAOS object class ID based on name.  """
+
+        # Default to OC_UNKNOWN (0), which will automatically select an object class.
+        if cid == "0":
+            return 0
+        if cid != "0":
+            return ObjClassID[cid].value
```
Comment (daltonbohning) [inline src/client/pydaos/pydaos_core.py:117]: This will simplify and resolve the pylint error
```suggestion
        if cid == "0":
            return 0
        return ObjClassID[cid].value
```

File: src/client/pydaos/pydaos_core.py [hunk 5]
```diff
@@ -107,12 +107,24 @@ def get(self, name):
...
+
+    def dict(self, name, v: dict = None, cid="0"):
```
Comment (daltonbohning) [inline src/client/pydaos/pydaos_core.py:119]: This adds a pylint error
`invalid-name, Argument name "v" doesn't conform to snake_case naming style`
This one is easy to fix, but some others like `hi`, `lo`, etc. are more difficult. To appease the pylint GHA check, you can add this to the top of the file (under the module docstring)
```
# pylint: disable=invalid-name
```

Comment (ashleypittman) [reply src/client/pydaos/pydaos_core.py:119]: For new code please use valid names and only suppress warnings for methods/attributes that are part of the published API.

Comment (ravalsam) [reply src/client/pydaos/pydaos_core.py:119]: dict() is an existing code, I just moved the function below.

Comment (ashleypittman) [reply src/client/pydaos/pydaos_core.py:119]: For what it's worth https://github.com/daos-stack/daos/pull/11825 fixes the pylint issues in this file, it might be an idea to cherry-pick the changes to this file in that PR into this PR.

Comment (ravalsam) [reply src/client/pydaos/pydaos_core.py:117]: Done

File: src/client/pydaos/pydaos_core.py [hunk 6]
```diff
@@ -22,6 +22,13 @@
     {key: value for key, value in list(pydaos_shim.__dict__.items())
      if key.startswith("OC_")})
 
+def _get_object_id(cid):
```
Comment (daltonbohning) [inline src/client/pydaos/pydaos_core.py:25]: flake wants 2 blank lines between top-level definitions
```suggestion

def _get_object_id(cid):
```

File: src/client/pydaos/pydaos_core.py [hunk 7]
```diff
@@ -22,6 +22,13 @@
...
+    """ Get the existing DAOS object class ID based on name.  """
+
+    # Default to OC_UNKNOWN (0), which will automatically select an object class.
+    if cid == "0":
+        return 0
+    return ObjClassID[cid].value
```
Comment (daltonbohning) [inline src/client/pydaos/pydaos_core.py:31]:
```suggestion
    return ObjClassID[cid].value

```

File: utils/node_local_test.py [hunk 8]
```diff
@@ -3988,6 +3988,76 @@ def test_pydaos_kv(server, conf):
...
+
+def test_pydaos_kv_obj_class(server, conf):
+    """Test the predefined object class works with KV"""
+    # pylint: disable=consider-using-with
+
+    pydaos_log_file = tempfile.NamedTemporaryFile(prefix='kv_objclass_pydaos_',
+                                                  suffix='.log',
+                                                  delete=False)
+
+    os.environ['D_LOG_FILE'] = pydaos_log_file.name
```
Comment (daltonbohning) [inline utils/node_local_test.py:4000]: Pre-existing in other places in this file, so out of scope - question for @ashleypittman If we just need the name, and `delete=False`, then shouldn't this work?
```suggestion
    with tempfile.NamedTemporaryFile(prefix='kv_objclass_pydaos_',
                                     suffix='.log',
                                     delete=False) as tmp_file:
        os.environ['D_LOG_FILE'] = tmp_file.name
```

**Review by daltonbohning — CHANGES_REQUESTED** [review]
Just the introduced flake errors

File: src/client/pydaos/pydaos_core.py [hunk 9]
```diff
@@ -124,13 +136,16 @@ def dict(self, name, v: dict = None):
 
         return dd
 
-    def array(self, name, v: list = None):
+    def array(self, name, v: list = None, cid="0"):
         # pylint: disable=unused-argument
         """ Create new DArray object """
 
+        # Get the existing class ID based on class name given. Default to 0
+        objId = _get_object_id(cid)
```
Comment (mchaarawi) [inline src/client/pydaos/pydaos_core.py:144]: just to verify this works if cid == 0?
is objID in this case (0)?
could be something like:
if (cid != "0")
  objId = _get_object_id(cid)
else
  objId = 0

File: utils/node_local_test.py [hunk 10]
```diff
@@ -3988,6 +3988,76 @@ def test_pydaos_kv(server, conf):
...
+    kv1 = container.dict('object1', {"Monday": "1"}, "OC_S2")
+    if len(kv1) != 1:
+        failed = True
+        print(f'Expected length of kv object is 1 but got {len(kv1)}')
+
+    # Write kv2 dictionary with OC_SX object type
+    kv2 = container.dict('object2', {"Monday": "1", "Tuesday": "2"}, "OC_SX")
+    if len(kv2) != 2:
+        failed = True
+        print(f'Expected length of kv object is 2 but got {len(kv2)}')
```
Comment (mchaarawi) [inline utils/node_local_test.py:4019]: maybe add 1 more dictionary that does not set the oclass. expected oclass should be S1 in this case.

File: src/client/pydaos/pydaos_core.py (hunk 9 above)
Comment (ravalsam) [reply src/client/pydaos/pydaos_core.py:144]: That check is in _get_object_id() function above.

File: utils/node_local_test.py (hunk 10 above)
Comment (ravalsam) [reply utils/node_local_test.py:4019]: will do

**Review by mchaarawi — APPROVED** [review]

**Comment by mchaarawi** [comment]
for future reference, you should add
Skip-func-test: true
since you are not modifying any core code or functional tests.

**Review by daltonbohning — DISMISSED** [review]

**Review by mchaarawi — DISMISSED** [review]

**Review by daltonbohning — APPROVED** [review]

File: src/client/pydaos/pydaos_shim.c
Comment (ashleypittman) [inline src/client/pydaos/pydaos_shim.c:1]: If you're changing any of the entry points in this file then I think we should be changing the PY_SHIM_MAGIC_NUMBER value as well.

File: src/client/pydaos/raw/README.md [hunk 11]
```diff
@@ -23,7 +23,6 @@ pool.connect(1 << 1)
 # query the pool
 pool_info = pool.pool_query()
 print("Pool has {0} storage targets".format(pool_info.pi_ntargets))
-print("Pool created with {0} permissions".format(pool_info.pi_mode))
```
Comment (ashleypittman) [inline src/client/pydaos/raw/README.md:26]: Is this change related?  pydaos.raw is not related to pydaos.

File: utils/node_local_test.py (hunk 8 above)
Comment (ashleypittman) [reply utils/node_local_test.py:4000]: Yes, that would be preferable, and then the "disable=consider-using-with" pragma can be removed as well.

It's one fewer local, one fewer suppression and the code is clearer in it's intent.

File: utils/node_local_test.py [hunk 12]
```diff
@@ -3988,6 +3988,77 @@ def test_pydaos_kv(server, conf):
...
+                                                  delete=False)
+
+    os.environ['D_LOG_FILE'] = pydaos_log_file.name
+    daos = import_daos(server, conf)
+
+    pool = server.get_test_pool_obj()
+
+    cont = create_cont(conf, pool, ctype="PYTHON")
+
+    container = daos.DCont(pool.uuid, cont.uuid)
```
Comment (ashleypittman) [inline utils/node_local_test.py:4007]: I think pool.label or pool.id() would get you a label here, but I'm not sure if daos.DCont() accepts labels.
```suggestion
    container = daos.DCont(pool.label, cont.uuid)
```

File: utils/node_local_test.py [hunk 13]
```diff
@@ -3988,6 +3988,77 @@ def test_pydaos_kv(server, conf):
...
+    pydaos_log_file = tempfile.NamedTemporaryFile(prefix='kv_objclass_pydaos_',
+                                                  suffix='.log',
+                                                  delete=False)
+
+    os.environ['D_LOG_FILE'] = pydaos_log_file.name
+    daos = import_daos(server, conf)
+
+    pool = server.get_test_pool_obj()
+
+    cont = create_cont(conf, pool, ctype="PYTHON")
```
Comment (ashleypittman) [inline utils/node_local_test.py:4005]: It would be good to pass in a label here.

File: utils/node_local_test.py [hunk 14]
```diff
@@ -3988,6 +3988,77 @@ def test_pydaos_kv(server, conf):
...
+
+    # Write kv2 dictionary without any object type,
+    # so in this case we have 4 targets so default object type should be S4
+    kv2 = container.dict('object2', {"Monday": "1", "Tuesday": "2"})
+    if len(kv2) != 2:
+        failed = True
+        print(f'Expected length of kv object is 2 but got {len(kv2)}')
+
+    # Run a command to list the objects
+    cmd = ['cont', 'list-objects', pool.uuid, cont.uuid]
```
Comment (ashleypittman) [inline utils/node_local_test.py:4023]: This should definitely be using labels.

File: utils/node_local_test.py [hunk 15]
```diff
@@ -3988,6 +3988,77 @@ def test_pydaos_kv(server, conf):
...
+    data = rc.json
+    assert data['status'] == 0, rc
+    assert data['error'] is None, rc
+    assert data['response'] is not None, rc
+
+    # Run a command to get the object layout
+    print('query the object layout')
+    actual_obj_layout = []
+    for obj in data['response']:
+        cmd = ['object', 'query', pool.uuid, cont.uuid, obj]
```
Comment (ashleypittman) [inline utils/node_local_test.py:4036]: As should this.

File: utils/node_local_test.py [hunk 16]
```diff
@@ -3988,6 +3988,77 @@ def test_pydaos_kv(server, conf):
...
+
+    # Verify the object has the correct layout used during kv dictionary creation.
+    expected_obj_layout = ['S2', 'S4']
+    for obj in expected_obj_layout:
+        if obj not in actual_obj_layout:
+            failed = True
+            print(f'Expected obj {obj} not found in all {actual_obj_layout}')
+
+    if failed:
+        print("Test step failed in test_pydaos_kv_obj_class()")
```
Comment (ashleypittman) [inline utils/node_local_test.py:4053]: This will only print to stdout but not actually fail anything.  You need something like the below to fail NLT.

`conf.wf.add_test_case('pydaos object test', failure='test failed')`

**Review by ashleypittman — COMMENTED** [review]
Given how much effort is going into removing the use of uuids from existing test cases I'd rather we added new ones using labels only.

File: utils/node_local_test.py (hunk 13 above)
Comment (ravalsam) [reply utils/node_local_test.py:4005]: I will update.

File: utils/node_local_test.py (hunk 16 above)
Comment (ravalsam) [reply utils/node_local_test.py:4053]: I will update

Comment (ashleypittman) [reply utils/node_local_test.py:4053]: Thinking about this more you want to create the test case anyway, rather than just on failure so you want something like:
`if failed:
   conf.wf.add_test_case('pydaos object test', failure='test failed')
else:
   conf.wf.add_test_case('pydaos object test')
`
and then you'll be able to tell from the test results that this code ran.

I note that the existing test_pydaos_kv() function does not do this, it long predates the junit reporting of results so it does warrant updating.  If you don't want to do it here that's fine, I can push a PR to bring that one up to spec once this PR is compete.

Comment (ravalsam) [reply utils/node_local_test.py:4053]: I have not updated all the review comments, so I can incorporate this as well. It's a good way to confirm. 
I used the dummy prints just to make sure it ran in Ci. 
I am not sure how this NLT results are populating in Jenkins, so anything gives the test confirmation will on console will be idle.

File: src/client/pydaos/__init__.py
Comment (ashleypittman) [inline src/client/pydaos/__init__.py:1]: Are you missing some updates from this file?  I thought you'd pulled in the pylint suppression changes?

Comment (ravalsam) [reply src/client/pydaos/__init__.py:1]: I fixed this at the last min and forgot to fix pylint issue. I will update the PR.

File: utils/node_local_test.py [hunk 17]
```diff
@@ -3988,6 +3988,79 @@ def test_pydaos_kv(server, conf):
...
+    expected_obj_layout = ['S2', 'S4']
+    for obj in expected_obj_layout:
+        if obj not in actual_obj_layout:
+            failed = True
+            print(f'Expected obj {obj} not found in all {actual_obj_layout}')
+
+    if failed:
+        conf.wf.add_test_case('pydaos kv object test', failure='test failed')
+    else:
+        conf.wf.add_test_case('pydaos kv object test')
```
Comment (ashleypittman) [inline utils/node_local_test.py:4055]: There's no stdout from this test but the result is registered at https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11952/30/testReport/NLT/core/Unit_Tests___NLT_on_EL_8___pydaos_kv_object_test/

**Review by ashleypittman — APPROVED** [review]

**Review by mchaarawi — APPROVED** [review]
